onnxruntime/core/util/thread_utils.cc

```diff
@@ -13,55 +13,66 @@ static inline std::vector<size_t> GenerateVectorOfN(size_t n) {
   std::iota(ret.begin(), ret.end(), 0);
   return ret;
-                }
+}
+
+struct NumCoresUtil {
 #ifdef _WIN32
-		// This function doesn't support systems with more than 64 logical processors
-		static std::vector<size_t> GetNumCpuCores() {
-			// Indeed 64 should be enough. However, it's harmless to have a little more.
-			SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-			DWORD returnLength = sizeof(buffer);
-			if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
-                          return GenerateVectorOfN(std::thread::hardware_concurrency());
-                        }
-                        std::vector<size_t> ret;
-			int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-			for (int i = 0; i != count; ++i) {
-				if (buffer[i].Relationship == RelationProcessorCore) {
-					ret.push_back(buffer[i].ProcessorMask);
-				}
-			}
-			if (ret.empty())
-                          return GenerateVectorOfN(std::thread::hardware_concurrency());
-                        return ret;
-                }
+  // This function doesn't support systems with more than 64 logical processors
+  static std::vector<size_t> GetNumCpuCores() {
+    // Indeed 64 should be enough. However, it's harmless to have a little more.
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
+    DWORD returnLength = sizeof(buffer);
+    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
+      return GenerateVectorOfN(std::thread::hardware_concurrency());
+    }
+    std::vector<size_t> ret;
+    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
+    for (int i = 0; i != count; ++i) {
+      if (buffer[i].Relationship == RelationProcessorCore) {
+        ret.push_back(buffer[i].ProcessorMask);
+      }
+    }
+    if (ret.empty())
+      return GenerateVectorOfN(std::thread::hardware_concurrency());
+    return ret;
+  }
 #else
-		static std::vector<size_t> GetNumCpuCores() {
-			return GenerateVectorOfN(std::thread::hardware_concurrency() / 2);
-		}
+  static std::vector<size_t> GetNumCpuCores() {
+    return GenerateVectorOfN(std::thread::hardware_concurrency() / 2);
+  }
 #endif
-		std::unique_ptr<ThreadPool> CreateThreadPool(Env* env, OrtThreadPoolParams options, Eigen::Allocator* allocator) {
-			if (options.thread_pool_size == 1)
-				return nullptr;
-			std::vector<size_t> cpu_list;
-			ThreadOptions to;
-			if (options.affinity_vec_len != 0) {
-				to.affinity.assign(options.affinity_vec, options.affinity_vec + options.affinity_vec_len);
-			}
-			if (options.thread_pool_size <= 0) {  // default
-				cpu_list = GetNumCpuCores();
-				if (cpu_list.empty() || cpu_list.size() == 1)
-					return nullptr;
-				options.thread_pool_size = static_cast<int>(cpu_list.size());
-				if (options.auto_set_affinity)
-					to.affinity = cpu_list;
-			}
+};
+#ifdef _OPENMP
+std::unique_ptr<ThreadPool>
+CreateThreadPool(Env*, OrtThreadPoolParams, Eigen::Allocator*) {
```

Comment:
Then you'll disable parallel executor too.

Reply:
Yeah, thinking about this a bit more I'm a bit hesitant to special case for the parallel executor (PE). Reason being how do we communicate to our users to configure the thread count without causing confusion and introducing 2 ways to control the thread count? We'll have to say something like if openmp is enabled, you can control num threads using the openmp env variables, unless you're using the parallel executor in which case use set session inter op API. Would PE work and be equivalently performant with just openmp?

Reply:
Ok, resolved as per discussion.